### PR TITLE
Removed the use of memory mutation for side-fork unwinds.

### PR DESCRIPTION
**Pull request by Giulio2002**

### CODE REVIEW

File: turbo/engineapi/fork_validator.go [hunk 1]
```diff
@@ -228,7 +229,6 @@ func (fv *ForkValidator) ValidatePayload(tx kv.RwTx, header *types.Header, body
 	currentHash := header.ParentHash
 	foundCanonical, criticalError = rawdb.IsCanonicalHash(tx, currentHash)
 	if criticalError != nil {
-		fmt.Println("critical")
 		return
```
Comment (mandrigin) [inline turbo/engineapi/fork_validator.go:232]: hiding errors is not good, at least let's add a warning log here

File: turbo/engineapi/fork_validator.go [hunk 2]
```diff
@@ -256,10 +256,30 @@ func (fv *ForkValidator) ValidatePayload(tx kv.RwTx, header *types.Header, body
 	if unwindPoint == fv.currentHeight {
 		unwindPoint = 0
 	}
-	// if it is not canonical we validate it in memory and discard it aferwards.
-	batch := memdb.NewMemoryBatch(tx)
-	defer batch.Close()
-	return fv.validateAndStorePayload(batch, header, body, unwindPoint, headersChain, bodiesChain)
+	status, latestValidHash, validationError, criticalError = fv.validateAndStorePayload(tx, header, body, unwindPoint, headersChain, bodiesChain)
+	if criticalError != nil {
+		return
```
Comment (mandrigin) [inline turbo/engineapi/fork_validator.go:261]: hiding errors is not good, at least let's add a warning log here

File: turbo/engineapi/fork_validator.go [hunk 3]
```diff
@@ -256,10 +256,30 @@ func (fv *ForkValidator) ValidatePayload(tx kv.RwTx, header *types.Header, body
...
+	}
+
+	if criticalError = tx.Reset(); criticalError != nil {
+		return
```
Comment (mandrigin) [inline turbo/engineapi/fork_validator.go:265]: hiding errors is not good, at least let's add a warning log here

File: turbo/engineapi/fork_validator.go [hunk 4]
```diff
@@ -256,10 +256,30 @@ func (fv *ForkValidator) ValidatePayload(tx kv.RwTx, header *types.Header, body
...
+	}
+	// We do this, in case we previously saved blocks and deleted them.
+	for i := len(headersChain) - 1; i >= 0; i-- {
+		var parentTd *big.Int
+		currentHeader := headersChain[i]
+		parentTd, criticalError = rawdb.ReadTd(tx, currentHeader.ParentHash, currentHeader.Number.Uint64()-1)
+		if criticalError != nil || parentTd == nil {
+			criticalError = fmt.Errorf("parent's total difficulty not found with error: %v", criticalError)
+			return
```
Comment (mandrigin) [inline turbo/engineapi/fork_validator.go:274]: hiding errors is not good, at least let's add a warning log here

File: turbo/engineapi/fork_validator.go (hunk 2 above)
Comment (Giulio2002) [reply turbo/engineapi/fork_validator.go:261]: We are not hiding errors we are still returning it

File: turbo/engineapi/fork_validator.go (hunk 3 above)
Comment (Giulio2002) [reply turbo/engineapi/fork_validator.go:265]: We are not hiding errors we are still returning it

File: turbo/engineapi/fork_validator.go (hunk 4 above)
Comment (Giulio2002) [reply turbo/engineapi/fork_validator.go:274]: We are not hiding errors we are still returning it

File: turbo/engineapi/fork_validator.go (hunk 1 above)
Comment (Giulio2002) [reply turbo/engineapi/fork_validator.go:232]: that is a debug println left there for accident

**Review by yperbasis — APPROVED** [review]

**Pull request merged**